### PR TITLE
Opal

### PR DESCRIPTION
**Pull request by webdev778**
## Changes
- Create GHA script to test Opal version of interscript
- Run all map tests in JS
- Fix up systems that work in Ruby but not in JS
## ScreenShots
image

## What should do after merge
- To generate the interscript javascript file
   bundle exec rake js
- Test 
   npm install
   npm test

### CODE REVIEW

File: lib/interscript.rb [hunk 1]
```diff
@@ -181,5 +195,14 @@ def up_case_around?(string, pos)
...
       before_uc || after_uc
     end
+
+    def mkregexp(regexpstring)
+      flags = 'u'
+      if regexpstring.include? "(?i)"
+        regexpstring = regexpstring.gsub("(?i)", '').gsub("(?-i)", '')
+        flags += 'i'
+      end
+      regexp = Regexp.new(regexpstring, flags)
```
Comment (ronaldtse) [inline lib/interscript.rb]: Maybe just remove `regexp = ` because Ruby returns the last evaluated line.

File: lib/interscript.rb [hunk 2]
```diff
@@ -133,9 +149,8 @@ def transliterate(system_code, string, maps={})
           # if more than one, choose the first one
           result = result[0] if result.is_a?(Array)
 
-          output = output[0, pos] +
-            add_separator(separator, pos, result) +
-            output[(pos + match[0].size)..-1]
+          output = output.sub_replace(pos, match[0].size,
+            add_separator(separator, pos, result))
```
Comment (ronaldtse) [inline lib/interscript.rb]: Agree with hound.

File: lib/interscript.rb [hunk 3]
```diff
@@ -121,7 +132,12 @@ def transliterate(system_code, string, maps={})
 
       mapping.rules.each do |r|
         if output
-          output = output.gsub(/#{r['pattern']}/u, r['result'])
+          if RUBY_ENGINE != "opal"
```
Comment (ronaldtse) [inline lib/interscript.rb]: Can we do:
```ruby
next unless output
if RUBY_ENGINE != "opal"
  output = output.gsub(/#{r["pattern"]}/u, r["result"])
else
...
```

File: .github/workflows/js.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    branches: [ master, opal ]
+
+jobs:
+  test:
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        node-version: [14.x]
```
Comment (ronaldtse) [inline .github/workflows/js.yaml]: Can we add 1-2 more node versions?

**Review by ronaldtse — CHANGES_REQUESTED** [review]
Nice work, thanks @webdev778 !

**Comment by ronaldtse** [comment]
Please also help fix the hound issues, thanks!

**Comment by ronaldtse** [comment]
@webdev778 I've also run into an issue with `npm test`:

```
npm test

> interscript@1.0.0 test /Users/me/src/interscript/interscript
> mocha

/Users/mulgogi/src/interscript/interscript/test/test.js:12
				var expected = test['expected']?.normalize();
				                                ^
SyntaxError: Unexpected token '.'
    at wrapSafe (internal/modules/cjs/loader.js:1072:16)
    at Module._compile (internal/modules/cjs/loader.js:1122:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
    at Module.load (internal/modules/cjs/loader.js:1002:32)
    at Function.Module._load (internal/modules/cjs/loader.js:901:14)
    at Module.require (internal/modules/cjs/loader.js:1044:19)
    at require (internal/modules/cjs/helpers.js:77:18)
```

It works if I change this line:
```
				var expected = test['expected']?.normalize();
```
to
```
				var expected = test['expected'].normalize();
```

**Review by webdev778 — COMMENTED** [review]
> @webdev778 I've also run into an issue with `npm test`:
> 
> ```
> npm test
> 
> > interscript@1.0.0 test /Users/me/src/interscript/interscript
> > mocha
> 
> /Users/mulgogi/src/interscript/interscript/test/test.js:12
> 				var expected = test['expected']?.normalize();
> 				                                ^
> SyntaxError: Unexpected token '.'
>     at wrapSafe (internal/modules/cjs/loader.js:1072:16)
>     at Module._compile (internal/modules/cjs/loader.js:1122:27)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
>     at Module.load (internal/modules/cjs/loader.js:1002:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:901:14)
>     at Module.require (internal/modules/cjs/loader.js:1044:19)
>     at require (internal/modules/cjs/helpers.js:77:18)
> ```
> 
> It works if I change this line:
> 
> ```
> 				var expected = test['expected']?.normalize();
> ```
> 
> to
> 
> ```
> 				var expected = test['expected'].normalize();
> ```

it's caused by node version as es6 syntax, I fixed it to be compatible with older versions of node.

File: lib/interscript.rb (hunk 1 above)
Comment (webdev778) [reply lib/interscript.rb]: fixed

File: lib/interscript.rb (hunk 2 above)
Comment (webdev778) [reply lib/interscript.rb]: fixed, it was marked as outdated, definitely I fixed before pushed

**Pull request merged**